StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp

```diff
@@ -0,0 +1,103 @@
+#include "UpgradeExplosiveAttack.h"
+#include "EngineMath.h"
+#include "ECS/Entity.h"
+#include "Player.h"
+#include "Components/ProjectileComponent.h"
+#include "EnemyFactory.h"
+#include "Misc/AssetLoader.h"
+#include "Renderer/Texture/Texture2DGUI.h"
+#include "ECS/SceneManager.h"
+#include "Renderer/Transform.h"
+#include "ECS/Components/TemporaryLifeComponent.h"
+#include "ECS/Components/TransformComponent.h"
+
+UpgradeExplosiveAttack::UpgradeExplosiveAttack(Entity* parent)
+	:Upgrade(parent)
+{
+	// name the upgrade!
+	SetName("UpgradeExplosiveAttack");
+	// set the upgrade type/types!
+	SetType(F_UpgradeType::RANGE);	// goes on the projectiles
+	// set the price of this upgrade
+	m_Price = 300;
+	m_StartingPrice = m_Price;
+
+	// percentage of damage done to steal as life
+	m_Radius = 3.0;
+	m_PercentageDamage = 0.5;
+
+	m_ImageName = "Explosion.png";
+}
+
+UpgradeExplosiveAttack::~UpgradeExplosiveAttack()
+{
+}
+
+void UpgradeExplosiveAttack::IncreaseLevel()
+{
+	m_Level++;
+	m_Radius = (5 * m_Level);
+	m_PercentageDamage = 0.4 + (0.1 * m_Level);
+	m_Price += m_StartingPrice;
+}
+
+void UpgradeExplosiveAttack::OnRangedHit(Entity* target)
+{
+	//Should only affect enemies. Only enemies should have AiComponent and therefore should be a good test
+	if (target->HasComponent<component::AiComponent>())
+	{
+		// get damage from projectile
+		int damage = m_pParentEntity->GetComponent<component::ProjectileComponent>()->GetDamage() * m_PercentageDamage;
+
+		std::vector<Entity*>* enemies = EnemyFactory::GetInstance().GetAllEnemies();
+		float3 hitTarget = target->GetComponent<component::TransformComponent>()->GetTransform()->GetPositionFloat3();
+		for (int i = 0; i < enemies->size(); i++)
+		{
+			if (enemies->at(i) != target)
+			{
+				if (m_Radius > (hitTarget - enemies->at(i)->GetComponent<component::TransformComponent>()->GetTransform()->GetPositionFloat3()).length())
+				{
+					enemies->at(i)->GetComponent<component::HealthComponent>()->TakeDamage(damage);
+				}
+			}
+		}
+
+		ParticleEffectSettings settings = {};
+		settings.maxParticleCount = 20;
+		settings.startValues.lifetime = 1.0;
+		settings.spawnInterval = 0.001;
+		settings.startValues.acceleration = { 0, 0, 0 };
```

Comment:
Would create a "base" ParticleEffectSettings on constructor, then in here modify that one with the only things that need to be modified, it is then later copied when creating the particleeffect.